File: pubsub/src/main/java/glitch/pubsub/object/adapters/ServerTimeAdapter.java
package glitch.pubsub.object.adapters;

import com.google.gson.*;
import java.lang.reflect.Type;
import java.time.Instant;

/**
 * @author Damian Staszewski [damian@stachuofficial.tv]
 * @version %I%, %G%
 * @since 1.0
 */
public class ServerTimeAdapter implements JsonSerializer<Instant>, JsonDeserializer<Instant> {

    @Override
    public Instant deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        String[] secdons = json.getAsString().split("\\.");
        long s = Long.parseLong(secdons[0]);
        long ns = Long.parseLong(secdons[1]) * 100;
        return Instant.ofEpochSecond(s, ns);
    }

    @Override
    public JsonElement serialize(Instant src, Type typeOfSrc, JsonSerializationContext context) {
        return new JsonPrimitive(src.getNano());
    }
}

File: pubsub/src/main/kotlin/io/glitchlib/pubsub/model/adapters/ServerTimeAdapter.kt
package io.glitchlib.pubsub.model.adapters

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.lang.reflect.Type
import java.time.Instant
import java.util.*

/**
 * @author Damian Staszewski [damian@stachuofficial.tv]
 * @version %I%, %G%
 * @since 1.0
 */
class ServerTimeAdapter : JsonSerializer<Date>, JsonDeserializer<Date> {

    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Date {
        val secdons = json.asString.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val s = secdons[0].toLong()
        val ns = secdons[1].toLong() * 100
        return Date.from(Instant.ofEpochSecond(s, ns))
    }

    override fun serialize(src: Date, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        return JsonPrimitive(src.toInstant().nano)
    }
}
